tests/apps/tests.py

```diff
@@ -520,3 +526,41 @@ def test_explicit_default_app_config_mismatch(self):
             with self.settings(INSTALLED_APPS=['apps.explicit_default_config_mismatch_app']):
                 config = apps.get_app_config('explicit_default_config_mismatch_app')
             self.assertIsInstance(config, ExplicitDefaultConfigMismatch)
+
+    def test_explicit_default_app_config_with_empty_apps(self):
+        """
+        Load an app that specifies a default AppConfig class in init and have
+        an empty apps file.
+        """
+        msg = (
+            "'apps.explicit_default_config_with_empty_apps' defines "
+            "default_app_config = 'apps.explicit_default_config_with_empty_apps."
+            "ExplicitDefaultConfigWithEmptyApps'. However, Django's "
+            "automatic detection did not found a configuration. You should "
+            "move the default config class to the apps submodule of your "
+            "application and, if this module defines several config classes, "
+            "mark the default one with default = True."
+        )
+        with self.assertRaisesMessage(RemovedInDjango41Warning, msg):
+            with self.settings(INSTALLED_APPS=['apps.explicit_default_config_with_empty_apps']):
+                config = apps.get_app_config('explicit_default_config_with_empty_apps')
+            self.assertIsInstance(config, ExplicitDefaultConfigWithEmptyApps)
+
+    def test_explicit_default_app_config_without_apps(self):
+        """
+        Load an app that specifies a default AppConfig class in init and do not
+        have an apps file.
+        """
+        msg = (
+            "'apps.explicit_default_config_without_apps' defines "
+            "default_app_config = 'apps.explicit_default_config_without_apps."
+            "ExplicitDefaultConfigWithoutApps'. However, Django's "
+            "automatic detection did not found a configuration. You should "
+            "move the default config class to the apps submodule of your "
+            "application and, if this module defines several config classes, "
+            "mark the default one with default = True."
+        )
+        with self.assertRaisesMessage(RemovedInDjango41Warning, msg):
+            with self.settings(INSTALLED_APPS=['apps.explicit_default_config_without_apps']):
+                config = apps.get_app_config('explicit_default_config_without_apps')
+            self.assertIsInstance(config, ExplicitDefaultConfigWithoutApps)
```

Comment:
This assertion is unreachable.

Reply:
It was copied from the other tests, they have the same problem of unreachable assertion.

Reply:
Yes, I've created PR #13289 to fix them.